File: coding/src/com/algos/core/li15_dp/revised/Partitions_L312_BurstBalloons.java
package  com.algos.core.li15_dp.revised;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Partitions_L312_BurstBalloons {

	public static void main(String[] args) {
		Partitions_L312_BurstBalloons obj = new Partitions_L312_BurstBalloons();

		int[] nums = new int[] { 3, 1, 5, 8 };
		System.out.println(obj.maxCoins2(nums));
		System.out.println(obj.maxCoins(nums));
	}

	// Problem: https://leetcode.com/problems/burst-balloons/
	// Idea: We take each position and try to burst the balloons on its left and
	// right before bursting it.
	// left sub problem and right sub problems become independent if we include the
	// current element also in its sub array.
	// Once left and right sub problems are solved, we are left with bursting last
	// balloon and last supporting balloons are at i-1 and j+1 positions
	public int maxCoins2(int[] nums) {
		int n = nums.length;
		List<Integer> list = Arrays.stream(nums).boxed().collect(Collectors.toList());
		list.add(0, 1);
		list.add(1);

		return maxCoins2(list, 1, n);		// n-1 shifted to n after inserting 1
	}

	private int maxCoins2(List<Integer> nums, int i, int j) {
		if (i > j) 
			return 0;

		int max = 0;
		for (int k = i; k <= j; k++) {
			int coins = nums.get(i - 1) * nums.get(k) * nums.get(j + 1)
					+ maxCoins2(nums, i, k - 1)
					+ maxCoins2(nums, k + 1, j);
			max = Math.max(max, coins);
		}
		return max;
	} 
	public int maxCoins(int[] nums) {
		int n = nums.length;
		List<Integer> list = Arrays.stream(nums).boxed().collect(Collectors.toList());
		list.add(0, 1);
		list.add(1);

		int[][] coins = new int[n + 2][n + 2];
		for(int i=n; i>=1; i--){
			for(int j=1; j<=n; j++){
				if(i > j) continue;

				int max = Integer.MIN_VALUE;
				for (int k = i; k <= j; k++) {
					int coin = list.get(i - 1) * list.get(k) * list.get(j + 1)
							+ coins[i][k - 1]
							+ coins[k + 1][j];
					max = Math.max(max, coin);
				}
				coins[i][j] = max;
			}
		}
		return coins[1][n];
	}
}

File: dsa/patterns/15_dp/python/partitions_l312_burst_balloons.py
# DP / Partition DP
#
# Problem: https://leetcode.com/problems/burst-balloons/
# Idea: We take each position and try to burst the balloons on its left and
# right before bursting it.
# Left sub problem and right sub problems become independent if we include the
# current element also in its sub array.
# Once left and right sub problems are solved, we are left with bursting last
# balloon and last supporting balloons are at i-1 and j+1 positions.


def maxCoins2(nums):
    n = len(nums)

    # add virtual balloons with value 1 at both ends
    arr = nums[:]
    arr.insert(0, 1)
    arr.append(1)

    return maxCoins2Helper(arr, 1, n)   # n-1 shifted to n after inserting 1


def maxCoins2Helper(nums, i, j):
    if i > j:
        return 0

    maximum = 0
    for k in range(i, j + 1):
        coins = (
            nums[i - 1] * nums[k] * nums[j + 1]
            + maxCoins2Helper(nums, i, k - 1)
            + maxCoins2Helper(nums, k + 1, j)
        )
        maximum = max(maximum, coins)

    return maximum


def maxCoins(nums):
    n = len(nums)

    # add virtual balloons
    arr = nums[:]
    arr.insert(0, 1)
    arr.append(1)

    dp = [[0] * (n + 2) for _ in range(n + 2)]

    for i in range(n, 0, -1):
        for j in range(1, n + 1):
            if i > j:
                continue

            maximum = float('-inf')
            for k in range(i, j + 1):
                coin = (
                    arr[i - 1] * arr[k] * arr[j + 1]
                    + dp[i][k - 1]
                    + dp[k + 1][j]
                )
                maximum = max(maximum, coin)

            dp[i][j] = maximum

    return dp[1][n]


# -------- Test --------
nums = [3, 1, 5, 8]
print(maxCoins2(nums))  # Expected: 167
print(maxCoins(nums))   # Expected: 167
